### PR TITLE
Docker Environment Setup Documentation

### PR DESCRIPTION
**Pull request by walshmm**
## Proposed changes

This PR adds documentation describing how to build and run QMCPACK on a docker container.  

## What type(s) of changes does this code introduce?
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
- Ubuntu 20.04

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

File: docs/running_docker.rst [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      cd build
+
+
+2. **Configure**:
+
+   .. code-block:: bash
+
+      cmake -GNinja \
+       -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+       -DMPI_C_COMPILER=mpicc -DMPI_CXX_COMPILER=mpicxx \
```
Comment (ye-luo) [inline docs/running_docker.rst:91]: Please do not use MPI_<LANGUAGE>_COMPILER, it is left anywhere in the current CI, please clean them up.

Comment (williamfgc) [reply docs/running_docker.rst:91]: @walshmm please change it to `-DCMAKE_C_COMPILER=mpicc -DMPI_CXX_COMPILER=mpicxx \` . Thanks.

**Review by prckent — CHANGES_REQUESTED** [review]
Please add a note about considering removing -DCMAKE_BUILD_TYPE=RelWithDebInfo for an optimized non-debug version.

**Comment by ye-luo** [comment]
> Please add a note about considering removing -DCMAKE_BUILD_TYPE=RelWithDebInfo for an optimized non-debug version.

The docker intro seems to explain how to reproduce the exact setting in the github-action CI, so if RelWithDebInfo is there, it needs to be mentioned here. So @walshmm please make them consistent.

File: docs/running_docker.rst (hunk 1 above)
Comment (walshmm) [reply docs/running_docker.rst:91]: @williamfgc for the CXX compiler did you mean `-DCMAKE_CXX_COMPILER=mpicxx` ?

Comment (williamfgc) [reply docs/running_docker.rst:91]: Yes, my bad, good catch. I need to stop doing C&P. Thanks!

**Comment by ye-luo** [comment]
After installing the docker engine. I need to do two additional steps.
1. start the engine service
2. add my username to the docker group, login and logout then hello world works.

**Comment by williamfgc** [comment]
@ye-luo just curious, did you have to restart your system after installation? Does the `docker` engine run automatically for you every time you restart?  Thanks for providing feedback, that would help improve the docs.

**Comment by prckent** [comment]
If we are confident being able to write "On first use you might need to..." would be helpful, but no need to mention platform specific and things that might change in docker imo. I didn't have to do anything for my existing install on mac.

**Comment by ye-luo** [comment]
I didn't reboot my machine. I'm testing under Windows -> WSL2 ubuntu 20.04 -> docker.

The build went well but all tests fail at
```
----------------------------------------------------------
--------------------------------------------------------------------------
mpiexec has detected an attempt to run as root.

Running as root is *strongly* discouraged as any mistake (e.g., in
defining TMPDIR) or bug can result in catastrophic damage to the OS
file system, leaving your system in an unusable state.

We strongly suggest that you run mpiexec as a non-root user.

You can override this protection by adding the --allow-run-as-root option
to the cmd line or by setting two environment variables in the following way:
the variable OMPI_ALLOW_RUN_AS_ROOT=1 to indicate the desire to override this
protection, and OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1 to confirm the choice and
add one more layer of certainty that you want to do so.
We reiterate our advice against doing so - please proceed at your own risk.
--------------------------------------------------------------------------
<end of output>
Test time =   0.00 sec
----------------------------------------------------------
Test Failed.
"deterministic-unit_test_wavefunction_trialwf" end time: Apr 29 21:20 UTC
"deterministic-unit_test_wavefunction_trialwf" time elapsed: 00:00:00
----------------------------------------------------------
```
can we make the user inside docker not root?

**Comment by williamfgc** [comment]
@ye-luo thanks for sharing. There is a caution box in the last sentence with a link to OpenMP docs. "user" should be the default non-root user. What's the `docker run` command you're using?

**Comment by ye-luo** [comment]
> 
> 
> @ye-luo thanks for sharing. There is a caution box in the last sentence with a link to OpenMP docs. "user" should be the default non-root user. What's the `docker run` command you're using?

I was following the documentation in this PR
```
docker run -u root -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
```

**Comment by williamfgc** [comment]
@ye-luo Thanks, that helps. Let us rework that section. The first `docker run` example shouldn't have the `-u` option.

**Comment by ye-luo** [comment]
I rebooted my windows and the docker service is not on at boot in WSL2.

**Comment by ye-luo** [comment]
@williamfgc please also update the docker image replacing libopenblas0 with libopenblas0-openmp

**Review by prckent — CHANGES_REQUESTED** [review]
"docker run -u root --env USER=`stat -c "%u" .` "
Can't we just have a default non root user? I thought the github ci worked that way.

s/QMCPack/QMCPACK/g   . Please always fully capitalize QMCPACK.

**Comment by walshmm** [comment]
> "docker run -u root --env USER=`stat -c "%u" .` "
> Can't we just have a default non root user? I thought the github ci worked that way.
> 
> s/QMCPack/QMCPACK/g . Please always fully capitalize QMCPACK.

The local user works fine for mimicking the flow seen in the ci-pipeline, as it pulls a fresh copy of the repo from git.

For development purposes, this will not work.

The reason for this is the local volume being mounted(your local repo with all your dev changes) is mounted to root by docker.  The default user does not have write permissions to this volume, which are necessary to build.  Since this volume isnt present in the initial docker image I'm not sure we can add permissions for it preemptively to the initial user, thus I need to add it upon creation of the container.  I found that simply creating a new user with the uid possessed by the user that initiated the container was a cleaner way of granting write permissions to this repo than messing with `chmod` or `chown`, which seem to have effects that permeated out of the container.

**Review by prckent — DISMISSED** [review]

**Comment by ye-luo** [comment]
This is your missing step
```
yeluo@ryzen-box:~$ sudo usermod -aG docker $USER
```
Once I confirm (maybe after logout/login, or even reboot)
```
yeluo@ryzen-box:~$ groups 
... docker 
```
No root works fine
```
docker run -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
```
I have no problem writing files

**Comment by ye-luo** [comment]
If root is used, all new files added are owned by "root" user when accessed from the host side. This looks scary.
If you do my way, inside the container, the new files are owned by "user", once I quit, files are owned by me "yeluo", very sane.

**Comment by walshmm** [comment]
> I have no problem writing files

I have followed your instructions listed here in a fresh pull of QMCPACK, I am still experiencing the same permission denied I was before during `cmake`
`CMake Error: Cannot open file for write: /home/user/build/CMakeCache.txt.tmp
CMake Error: : System Error: Permission denied
`
image
Tested on Ubuntu 20.04

> all new files added are owned by "root" user when accessed from the host side.

root is not used to generate the files(unless you launched docker with root), just a user.  The instructions include a command to swap over to an instance of bash ran with the new non root user.  I have just tested it now to confirm.

image

In docker `ls -l` says these files belong to user `local` and outside of docker they belong to my user `wqp` with a uid of `27675`

**Comment by ye-luo** [comment]
@walshmm what I meant is I don't see why `-u root --env USER=`stat -c "%u" .` we need this fancy gymnastics.

**Comment by ye-luo** [comment]
```
yeluo@ryzen-box:~$ groups
yeluo adm cdrom sudo dip plugdev lpadmin lxd sambashare docker software
yeluo@ryzen-box:~$ docker run -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

user@cb072c0ad35d:~$ mkdir build
mkdir: cannot create directory 'build': File exists
user@cb072c0ad35d:~$ exit
exit
yeluo@ryzen-box:~$ groups
yeluo adm cdrom sudo dip plugdev lpadmin lxd sambashare docker software
yeluo@ryzen-box:~$ docker run -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

user@10be02b87033:~$ mkdir build_docker_real
user@10be02b87033:~$ cd build_docker_real/
user@10be02b87033:~/build_docker_real$ cmake ..
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
^C
user@10be02b87033:~/build_docker_real$ exit
exit
yeluo@ryzen-box:~$ cd opt/qmcpack/
yeluo@ryzen-box:~/opt/qmcpack$ ls -lrt build_docker_real/
total 12
drwxr-xr-x 4 yeluo yeluo 4096 Apr 30 15:17 CMakeFiles
drwxr-xr-x 3 yeluo yeluo 4096 Apr 30 15:17 Testing
-rw-r--r-- 1 yeluo yeluo 2569 Apr 30 15:17 DartConfiguration.tcl
```

**Comment by walshmm** [comment]
@ye-luo 

```
wqp@LAP123315:~/git/qmcpack/build$ groups
users sudo docker
wqp@LAP123315:~/git/qmcpack/build$ docker run -v /home/wqp/git/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

user@c9e2a50aa86d:~$ mkdir build_docker_real
mkdir: cannot create directory 'build_docker_real': Permission denied
user@c9e2a50aa86d:~$ 
```

**Comment by walshmm** [comment]
@ye-luo 

If you run the docker container with the flags
 ```
-u root --env USER=`stat -c "%u" .`
```
 and input the command `sudo useradd -u $USER local && sudo -u local bash && exit` before building, what are the results of `ls -l` ?  You had mentioned they were owned by root, however I did not experience this, does that still persist after the above commands?

**Comment by ye-luo** [comment]
```
yeluo@ryzen-box:~/opt/qmcpack$ docker run -u root --env USER=`stat -c "%u" .`  -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
root@f9a2251eae19:/home/user# touch abc
root@f9a2251eae19:/home/user# exit
exit
yeluo@ryzen-box:~/opt/qmcpack$ ls -l
total 340
-rw-r--r--  1 root  root       0 Apr 30 16:11 abc
-rwxrwxr-x  1 yeluo yeluo 120192 Apr 10 16:20 a.out
```
I exercised this on both my WSL2 ubuntu 20.04 and my workstation. The same behavior. I don't thing it is right to fiddle with root at the first place. I don't know why your ubuntu is so different from mine. 

Also tried your lines
```
yeluo@ryzen-box:~/opt/qmcpack$ docker run -u root --env USER=`stat -c "%u" .`  -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
root@52013ea6dbb1:/home/user# sudo useradd -u $USER local && sudo -u local bash && exit 
useradd: UID 1000 is not unique
root@52013ea6dbb1:/home/user# touch abcde
root@52013ea6dbb1:/home/user# exit
exit
yeluo@ryzen-box:~/opt/qmcpack$ ls -l
total 340
-rw-r--r--  1 root  root       0 Apr 30 16:15 abcde
-rwxrwxr-x  1 yeluo yeluo 120192 Apr 10 16:20 a.out
```

**Comment by walshmm** [comment]
> ping @walshmm

Thank you. Currently spinning up a vm to confirm whether or not its just an issue with my local machine.  Will adjust docs accordingly.

**Comment by walshmm** [comment]
@ye-luo I am still experiencing the same issue on the vm.  Just to be sure, what OS are you using?

I will attempt to find another workaround that doesnt involve docker root.

**Comment by ye-luo** [comment]
> @ye-luo I am still experiencing the same issue on the vm. Just to be sure, what OS are you using?
> 
> I will attempt to find another workaround that doesnt involve docker root.

I was running Ubuntu 20.04 on my workstation and on my laptop under WSL2.

**Comment by walshmm** [comment]
```
docker run -u $(id -u `stat -c "%U" .`):$(id -g `stat -c "%G" .`) -v /home/wqp/git/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
```

@ye-luo  will this work for you?  It shouldn't touch root but instead directly interface with with the container as the current user/group

**Comment by ye-luo** [comment]
Looks good now.
```
yeluo@ryzen-box:~/opt/qmcpack$ docker run -u $(id -u `stat -c "%U" .`):$(id -g `stat -c "%G" .`) -v /home/yeluo/opt/qmcpack:/home/user -it williamfgc/qmcpack-ci:ubuntu20-openmpi /bin/bash
user@d0e09bf5e8fa:~$ touch aa
user@d0e09bf5e8fa:~$ ls -l aa
-rw-r--r-- 1 user user 0 May  7 17:43 aa
user@d0e09bf5e8fa:~$ exit
exit
yeluo@ryzen-box:~/opt/qmcpack$ ls -l aa
-rw-r--r-- 1 yeluo yeluo 0 May  7 12:43 aa
```

**Comment by walshmm** [comment]
> although I don't understand why it is needed to query user via filesystem instead of $USER.

There's no specific reason for this, just the first way I found to query both user and group.

Alright, so if everything looks good on that end I will change the docs to reflect this.

**Comment by ye-luo** [comment]
Okay to test

**Pull request merged**